src/SourceBuild/content/eng/tools/ExclusionsLibrary/ExclusionsStorage.cs

```diff
@@ -0,0 +1,113 @@
+
+namespace ExclusionsLibrary;
+
+internal class ExclusionsStorage
+{
+    /// <summary>
+    /// Storage for exclusions.
+    /// Key: file path.
+    /// Value: list of exclusions for the file.
+    /// </summary>
+    private Dictionary<string, List<Exclusion>> _storage = new();
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="ExclusionsStorage"/> class.
+    /// </summary>
+    public ExclusionsStorage() { }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="ExclusionsStorage"/> class.
+    /// <param name="other">The storage to copy.</param>
+    /// </summary>
+    public ExclusionsStorage(ExclusionsStorage other)
+    {
+        foreach (string file in other._storage.Keys)
+        {
+            _storage[file] = other._storage[file].Select(e => new Exclusion(e)).ToList();
+        }
+    }
+
+    /// <summary>
+    /// Gets all files in the storage.
+    /// </summary>
+    public IEnumerable<string> GetFiles() => _storage.Keys;
+
+    /// <summary>
+    /// Adds an exclusion to the storage.
+    /// <param name="file">The file to add the exclusion to.</param>
+    /// <param name="exclusion">The exclusion to add.</param>
+    /// </summary>
+    public void Add(string file, Exclusion exclusion)
+    {
+        if (!_storage.ContainsKey(file))
+        {
+            _storage[file] = new List<Exclusion>();
+        }
+        _storage[file].Add(exclusion);
+    }
+
+    /// <summary>
+    /// Removes a suffix from an exclusion in the storage.
+    /// If there are no suffixes left, the exclusion will be removed.
+    /// <param name="file">The file for the exclusion.</param>
+    /// <param name="pattern">The pattern to look for.</param>
+    /// <param name="suffix">The suffix to remove.</param>
+    /// </summary>
+    public void Remove(string file, string pattern, string? suffix)
+    {
+        if (_storage.ContainsKey(file))
+        {
+            Exclusion? exclusion = _storage[file].FirstOrDefault(e => e.Pattern == pattern);
+            if (exclusion is not null)
+            {
+                _storage[file].Remove(exclusion);
+                exclusion.Suffixes.Remove(suffix);
+                if (exclusion.Suffixes.Count > 0)
+                {
+                    _storage[file].Add(exclusion);
+                }
+            }
+        }
+    }
+
+    /// <summary>
+    /// Checks if the storage contains a file.
+    /// <param name="file">The file to check for.</param>
+    /// </summary>
+    public bool ContainsFile(string file) => _storage.ContainsKey(file);
+
+    /// <summary>
+    /// Gets an exclusion from the storage.
+    /// <param name="file">The file to get the exclusion from.</param>
+    /// <param name="pattern">The pattern to look for.</param>
+    /// </summary>
+    public Exclusion? GetExclusion(string file, string pattern) =>
+        _storage.ContainsKey(file) ? _storage[file].FirstOrDefault(e => e.Pattern == pattern) : null;
```

Comment:
This isn't taking advantage of Dictionary lookups. It's first doing a hash lookup using `ContainsKey` and if that returns true, then it does an enumeration of the dictionary to find the pattern when using `FirstOrDefault`. Instead, it should just use one call to the `Dictionary<K, V>.TryGetValue` method.